File: 739-daily-temperatures/daily-temperatures.cpp
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        //O(1) space complexity
        int n = temperatures.size();
        vector<int> sol (n,0);
        int hottest = -1;
        for(int i = n-1; i >=0 ; i-- )
        {
            if(temperatures[i]>= hottest)
            {
                hottest = temperatures[i];
                continue;
            }
        
            int day=1;
            
            
            while(temperatures[i+day] <= temperatures[i])
            {
                day+=sol[i+day];
            }
            sol[i]=day;
            
        }
        return sol;
        

    }
};

File: 739-daily-temperatures/daily-temperatures.py
class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        stack = []
        sol = [0]*len(temperatures)
        for i in range(len(temperatures)-1,-1,-1):
            while len(stack)>0 and stack[-1][0]<=temperatures[i]:
                stack.pop()
            
            if len(stack):
                sol[i] = stack[-1][1]-i
            
            stack.append([temperatures[i],i])
        return sol